rfcs/0004-pablo-distribution.adoc

```diff
@@ -0,0 +1,480 @@
+
+x _n = P_(current) . s_n / (sum s_n)
+
+"**Therefore existing staker receives the same reward as before.**"
+
+
+"To compensate for this new adjustment, a reduction " d_(n + 1) " of reward for each staker needs to be tracked,"
+
+x_(n+1) = P_(current) . s_(n+1) / (sum s_(n+1)) - d_(n+1)
+
+"In general"
+
+d_n = "Any reduction in reward including already claimed"
+
+------------------------------------------------
+
+"When adding a new reward to the pool the calculations remain the same other than,"
+
+P_(current) = P + rt + P_(added)
+
+"Since already claimed rewards("d_n") are tracked for each staker, "
+
+"they can always claim the new reward share from " P_(added) " later."
+
+++++
+
+As this method uses a reward pooling based approach to calculate the rewards for each staker out of it on-demand, rest of the document refers to this as the "reward pooling(*RP*) based approach".
+
+==== Data Structures
+
+Staking rewards pallet already uses the following fNFT data structure,
+
+[source,rust]
+----
+include::../frame/composable-traits/src/staking_rewards.rs[lines=103..122]
+----
+This data structure can be modified to support the RP approach by adding/adjusting/removing some fields.
+
+[source,rust]
+----
+pub struct StakingNFT<AccountId, AssetId, Balance, Epoch, Rewards> {
+    /// Reward Pool ID from which pool to allocate rewards for this
+    pub reward_pool_id: ID,
+
+    /// The staked asset.
+    pub asset: AssetId,
+    /// The original stake this NFT was minted for.
+    /// Used for calculating the pool share.
+    pub stake: Balance,
+
+    /// New: Pool share received for this NFT
+    pub share: Balance,
+
+    /// New: reduced rewards for the fNFT (d_n)
+    pub reduction: Balance,
+
+    /// The reward epoch at which this NFT will start yielding rewards. This is no longer needed.
+    /// pub reward_epoch_start: Epoch,
+
+    /// List of reward asset/pending rewards. This is no longer needed as pending rewards are pooled
+    /// pub pending_rewards: Rewards,
+
+    /// The date at which this NFT was minted.
+    pub lock_date: Timestamp,
+
+    /// The duration for which this NFT stake was locked.
+    pub lock_duration: DurationSeconds,
+
+    /// The penalty applied if a staker unstake before the end date.
+    pub early_unstake_penalty: Penalty<AccountId>,
+
+    /// The reward multiplier. Can remove this and calculate the extra shares a user gets based on the multiplier
+    /// pub reward_multiplier: Perbill,
+}
+----
+
+Now in order to allow redeeming the above fNFT, following data structures needs to be tracked,
+
+
+
+[#_rewardaccumulationhook]
+==== [Needs modification] RewardAccumulationHook
+
+Following algorithm is to added as part of the existing https://github.com/ComposableFi/composable/blob/main/frame/staking-rewards/src/lib.rs#L363[block hook] in staking rewards pallet. As it is only accumulating new rewards for an upcoming epoch, the code is proposed to be run inside a new state `State:AccumulatingRewards`.
```

Comment:
i think pablo can have its own loop calling stakin-rewards, no need to modify staking-rewards. so there can be addtional pallet for configuration of lp/pablo/pica prices relation.

Reply:
Note that this part is outdated.